content/client-lib-development-guide/features.textile

```diff
@@ -267,7 +268,19 @@ h3(#rest-channel). Channel
-** @(RSL6b)@ If, for example, incompatible encryption details are provided or invalid Base64 is detected in the message payload, an error message will be sent to the logger, but the message will still be delivered with last successful decoding and the @encoding@ field. For example, if a message had a decoding of "utf-8/cipher+aes-128-cbc/base64", and the payload was successfully Base64 decoded but the payload could not be encrypted because the @CipherParam@ details were not configured, the message would be delivered with a binary payload and an @encoding@ with the value "utf-8/cipher+aes-128-cbc"
+** @(RSL6b)@ If, for example, incompatible encryption details are provided or invalid Base64 is detected in the message payload, an error message will be sent to the logger, but the message will still be delivered with last successful decoding and the @encoding@ field. For example, if a message had a decoding of "utf-8/cipher+aes-128-cbc/base64", and the payload was successfully Base64 decoded but the payload could not be encrypted because the @CipherParam@ details were not configured, the message would be delivered with a binary payload and an @encoding@ with the value "utf-8/cipher+aes-128-cbc. Additional steps need to be taken if decoding failed on "vcdiff" encoding, see @(RSL6c)@ 
+** @(RSL6c)@ If decoding fails on "vcdiff" encoding that means message delta application failed.Since applying a delta depends on the previous message being present a failure in the delta application on the client side makes it impossible to apply any subsequent deltas received. That would render the channel unusable so a full message needs to be requested from Ably (the server) to replace the failed delta message. That is done by:
+*** @(RSL6c1)@ Following the procedure described in @(RTL4c)@ under the caveats of @(RTL4d)@, @(RTL4e)@ and @(RTL4f)@. The ProtocolMessage.channelSerial of the ATTACH ProtocolMessage sent as per @(RTL4c)@ should be set to the channel serial of the message which "vcdiff" decoding failed.
+*** @(RSL6c2)@ Any subsequent "vcdiff" decoding failures as per "RSL6c":#RSL6c while the recovery procedure from a previous failure as per "RSL6c1":#RSL6c1 is still in progress, i.e. the channel is in ATTACHING state, should not trigger a new recovery procedure as per "RSL6c1":#RSL6c1.
+** @(RSL6d)@ The last message on each channel has to be stored at all times since it might be needed to decode the "vcdiff" encoding. The last message bookkeping rules are:
+*** @(RSL6d1)@ Any transport layer introducing encoding (e.g. base64 introduced by sending binary on text transport) should be decoded before storing the last message
+*** @(RSL6d2)@ If "vcdiff" encoding is present the direct output of decoding the "vcdiff" encoding is stored as a last message, no matter that more encodings might be present
+
+h3(#pluggable-codecs). Pluggable Codecs
```

Comment:
Pending agreement as to whether this is a 'codec' or a 'plugin'.

see https://github.com/ably/ably-java/pull/532#discussion_r350645112

Your thoughts, please, @paddybyers and @mattheworiordan.

Reply:
Could think of no stopper so made this generic plugins.